qutebrowser/misc/sql.py

```diff
@@ -168,82 +171,212 @@ def raise_sqlite_error(msg, error):
+        self._user_version = self._USER_VERSION
+
+    def close(self) -> None:
+        """Close the SQL connection."""
+        database = self.qSqlDatabase()
+        database.close()
+        sip.delete(database)
+        QSqlDatabase.removeDatabase(self._path)
+
+    def transaction(self) -> 'Transaction':
+        """Return a Transaction object linked to this Database."""
+        return Transaction(self)
+
+    @classmethod
+    def version(cls) -> str:
+        """Return the sqlite version string."""
+        try:
+            in_memory_db = cls(':memory:')
+            version = in_memory_db.query("select sqlite_version()").run().value()
+            in_memory_db.close()
+            return version
+        except KnownError as e:
+            return f'UNAVAILABLE ({e})'
+
+
+class Transaction(contextlib.AbstractContextManager[None]):
+
+    """A Database transaction that can be used as a context manager."""
+
+    def __init__(self, database: Database):
+        self._database = database
+
+    def __enter__(self) -> None:
+        log.sql.debug('Starting a transaction')
+        db = self._database.qSqlDatabase()
+        ok = db.transaction()
+        if not ok:
+            error = db.lastError()
+            msg = f'Failed to start a transaction: "{error.text()}"'
+            raise_sqlite_error(msg, error)
+
+    def __exit__(self,
+                 _exc_type: Optional[Type[BaseException]],
+                 exc_val: Optional[BaseException],
+                 _exc_tb: Optional[types.TracebackType]) -> None:
+        db = self._database.qSqlDatabase()
+        if exc_val:
+            log.sql.debug('Rolling back a transaction')
+            db.rollback()
+        else:
+            log.sql.debug('Committing a transaction')
+            ok = db.commit()
+            if not ok:
+                error = db.lastError()
+                msg = f'Failed to commit a transaction: "{error.text()}"'
+                raise_sqlite_error(msg, error)
 
 
 class Query:
 
     """A prepared SQL query."""
 
-    def __init__(self, querystr, forward_only=True):
+    def __init__(self, database: Database, querystr: str,
+                 forward_only: bool = True):
         """Prepare a new SQL query.
 
         Args:
+            database: The Database object on which to operate.
             querystr: String to prepare query from.
             forward_only: Optimization for queries that will only step forward.
                           Must be false for completion queries.
         """
-        self.query = QSqlQuery(QSqlDatabase.database())
+        self._database = database
+        self.query = QSqlQuery(database.qSqlDatabase())
 
         log.sql.vdebug(f'Preparing: {querystr}')  # type: ignore[attr-defined]
         ok = self.query.prepare(querystr)
         self._check_ok('prepare', ok)
         self.query.setForwardOnly(forward_only)
 
-    def __iter__(self):
+    def __iter__(self) -> Iterator[Any]:  # FIXME: Iterator of what?
```

Comment:
`Iterator[collections.namedtuple]` probably? Not sure if this works, though.

Reply:
I'll try, thanks.

Reply:
Hmm, no, it's not a type according to mypy. typing.NamedTuple also wasn't accepted. I'll look into it again later.

Reply:
Ah, yes, those are type factories so to speak - see the [mypy docs](https://mypy.readthedocs.io/en/stable/kinds_of_types.html#named-tuples) for now they're normally used.

However, we're dynamically generating the `ResultRow` namedtuple (`rowtype`) based on what's in the database. Given that we can't really know statically which attributes it will have, I think `Any` is the correct thing to use.